lib/algos/list/tools.js

```diff
@@ -31,11 +32,49 @@ function inc(str) {
             String.fromCharCode(str.charCodeAt(str.length - 1) + 1)) : str;
 }
 
+/*
+ * Compares two Strings, s1 and s2, using UTF-8 lexicographic ordering.
+ * @function
+ * @param {String} s1 the first string to compare
+ * @param {String} s2 the second string to compare
+ * @return {number} -1, 0, or 1 if s1 is less than, equal or greater than s2 respectively
+ */
+function utf8Compare(s1, s2) {
+    const l1 = s1.length;
+    const l2 = s2.length;
+    const l = Math.min(l1, l2)
+
+    for (let i = 0; i < l; i++) {
```

Comment:
Side note, I was trying to find a way to convert this to more functional javascript with .map and .filter but JS doesn't seem to have a nice zip method like python does to loop over two iterables. Something along the lines of
```javascript
const c1 = Array.from(s1).map(c => c.codePointAt(0)).filter(c => c <= 0xffff);
const c2 = Array.from(s2).map(c => c.codePointAt(0)).filter(c => c <= 0xffff);
// but you would still have to loop over the indecies anyway
// plus having to go from string to array is kinda lame
```

Sure wish you could map over a string in javascript... seems a bit weird you can not.
Anyway, this was just me messing around.

Reply:
I tried to make this thing a bit more elegant as well, but I reckoned it could be rather performance critical so I went for an imperative, zero-allocation style in the end. Not pretty, but it does the job.

Reply:
Yea was thinking the same